src/mcts/search.cc

```diff
@@ -835,104 +866,96 @@ int SearchWorker::PrefetchIntoCache(Node* node, int budget) {
-    Node* node = node_to_process.node;
-    if (!node_to_process.nn_queried) {
-      // Terminal nodes don't involve the neural NetworkComputation, nor do
-      // they require any further processing after value retrieval.
-      node_to_process.v = node->GetQ();
-      continue;
-    }
-    // For NN results, we need to populate policy as well as value.
-    // First the value...
-    node_to_process.v = -computation_->GetQVal(idx_in_computation);
-    // ...and secondly, the policy data.
-    float total = 0.0;
-    for (auto edge : node->Edges()) {
-      float p = computation_->GetPVal(idx_in_computation,
-                                      edge.GetMove().as_nn_index());
-      if (search_->kPolicySoftmaxTemp != 1.0f) {
-        p = pow(p, 1 / search_->kPolicySoftmaxTemp);
-      }
-      total += p;
-      edge.edge()->SetP(p);
-    }
-    // Normalize P values to add up to 1.0.
-    if (total > 0.0f) {
-      float scale = 1.0f / total;
-      for (auto edge : node->Edges()) edge.edge()->SetP(edge.GetP() * scale);
-    }
-    // Add Dirichlet noise if enabled and at root.
-    if (search_->kNoise && node == search_->root_node_) {
-      ApplyDirichletNoise(node, 0.25, 0.3);
+  for (auto& node_to_process : minibatch_) {
+    FetchSingleNodeResult(&node_to_process, idx_in_computation);
+    if (node_to_process.nn_queried) ++idx_in_computation;
+  }
+}
+
+void SearchWorker::FetchSingleNodeResult(NodeToProcess* node_to_process,
+                                         int idx_in_computation) {
+  Node* node = node_to_process->node;
+  if (!node_to_process->nn_queried) {
+    // Terminal nodes don't involve the neural NetworkComputation, nor do
+    // they require any further processing after value retrieval.
+    node_to_process->v = node->GetQ();
+    return;
+  }
+  // For NN results, we need to populate policy as well as value.
+  // First the value...
+  node_to_process->v = -computation_->GetQVal(idx_in_computation);
+  // ...and secondly, the policy data.
+  float total = 0.0;
+  for (auto edge : node->Edges()) {
+    float p =
+        computation_->GetPVal(idx_in_computation, edge.GetMove().as_nn_index());
+    if (search_->kPolicySoftmaxTemp != 1.0f) {
+      p = pow(p, 1 / search_->kPolicySoftmaxTemp);
     }
-    ++idx_in_computation;
+    total += p;
+    edge.edge()->SetP(p);
+  }
+  // Normalize P values to add up to 1.0.
+  if (total > 0.0f) {
+    float scale = 1.0f / total;
+    for (auto edge : node->Edges()) edge.edge()->SetP(edge.GetP() * scale);
+  }
+  // Add Dirichlet noise if enabled and at root.
+  if (search_->kNoise && node == search_->root_node_) {
+    ApplyDirichletNoise(node, 0.25, 0.3);
   }
 }
 
 // 6. Propagate the new nodes' information to all their parents in the tree.
 // ~~~~~~~~~~~~~~
 void SearchWorker::DoBackupUpdate() {
-  // Initialize playout and depth counters.
-  uint16_t max_depth_batch = 0;
-  uint32_t cum_depth_batch = 0;
-  uint16_t playouts_batch = 0;
```

Comment:
This PR now has the side effect of changing the depth updates from per batch to per playout, but in the short term at least that's not a problem

Reply:
I'm not convinced that doing batch updates is faster than direct updates, as we hold exclusive mutex at that time anyway.

Reply:
Yes currently its a wash. I was hoping for a mutex reorg to gain some traction, but maybe there are more pressing things?!